### PR TITLE
Add active directory support to azure client runtime

### PR DESCRIPTION
**Pull request by markcowl**
- Includes build changes
- Includes design review feedback

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@markcowl__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (markcowl). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, or work for Microsoft Open Technologies, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        public string TokenType
+        {
+            get { return _type; }
+        }
+
+        /// <summary>
+        /// Gets an access token from the token cache or from AD authentication endpoint. 
+        /// Attempts to refresh the access token if it has expired.
+        /// </summary>
+        public virtual async Task<string> GetAccessTokenAsync(CancellationToken cancellationToken)
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs]: remove the "Access" as well?

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                        environment.ValidateAuthority,
+                        cache);
+            }
+            catch (AdalException authenticationException)
+            {
+                throw new AuthenticationException(Resources.ErrorCreatingAuthenticationContext, authenticationException);
+            }
+        }
+        /// <summary>
+        /// Set the ActiveDirectory authentication properties for this user
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs]: This comments doesn't match the method name of "Validate", should we rename the method to be "SetAuthenticationResult"?

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryUserTokenProvider.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        /// Create a token provider using Active Directory user credentials (UPN). 
+        /// Authentication occurs using the given username and password, with no user prompt.
+        /// </summary>
+        /// <param name="clientId">The client id for thsi application.</param>
+        /// <param name="domain">The domain or tenant id contianing the resources to manage.</param>
+        /// <param name="username">The username to use for authentication.</param>
+        /// <param name="password">The secret password associated with thsi user.</param>
+        /// <param name="environment">The azure environment to manage resources in.</param>
+        /// <param name="cache">The token cache to use during authentication.</param>
+        internal ActiveDirectoryUserTokenProvider(string clientId, string domain, string username, string password, AzureEnvironment environment, TokenCache cache)
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryUserTokenProvider.cs]: since the class is internal already, we can just set "public" if there are other types in the assemblies using it, or just "private"

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/UserAccessTokenCredentials.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+using System;
+using Microsoft.Rest;
+
+namespace Microsoft.Azure.Authentication
+{
+    /// <summary>
+    /// Credentials class for authenticating using a Microsoft ID or Organizational Id
+    /// </summary>
+    public class UserAccessTokenCredentials : TokenCredentials
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/UserAccessTokenCredentials.cs]: Remove the "Access"?

File: ClientRuntimes/CSharp/ClientRuntime.Azure/ClientRuntime.Azure.csproj [hunk 5]
```diff
@@ -19,12 +19,6 @@
   <PropertyGroup Condition="'$(Configuration)' == 'Net45-Release'">
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/ClientRuntime.Azure.csproj]: remove this

File: ClientRuntimes/CSharp/ClientRuntime.Azure/ClientRuntime.Azure.csproj [hunk 6]
```diff
@@ -19,12 +19,6 @@
...
     <DocumentationFile>bin\Net45-Release\Microsoft.Rest.ClientRuntime.Azure.xml</DocumentationFile>
   </PropertyGroup>
-  <PropertyGroup Condition="'$(Configuration)' == 'Net40-Debug'">
-    <DocumentationFile>bin\Net40-Debug\Microsoft.Rest.ClientRuntime.Azure.xml</DocumentationFile>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(Configuration)' == 'Net40-Release'">
-    <DocumentationFile>bin\Net40-Release\Microsoft.Rest.ClientRuntime.Azure.xml</DocumentationFile>
-  </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)' == 'Portable-Debug'">
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/ClientRuntime.Azure.csproj]: and this, as the common setting target cover it

File: ClientRuntimes/CSharp/ClientRuntime.Azure/ClientRuntime.Azure.csproj [hunk 7]
```diff
@@ -64,6 +56,14 @@
       <DependentUpon>Resources.resx</DependentUpon>
     </Compile>
   </ItemGroup>
+  <ItemGroup Condition=" '$(LibraryFxTarget)' == 'portable' or'$(LibraryFxTarget)' == 'net45' ">
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/ClientRuntime.Azure.csproj]: remove the condition, since we do not support net40 anymore

File: ClientRuntimes/CSharp/ClientRuntime.Azure/JsonSerializerExtensions.cs [hunk 8]
```diff
@@ -24,8 +24,8 @@ public static class JsonSerializerExtensions
                 throw new ArgumentNullException("serializer");
             }
             JsonSerializer newSerializer = new JsonSerializer();
-            PropertyInfo[] properties = typeof(JsonSerializer).GetProperties();
-            foreach (var property in properties.Where(p => p.GetSetMethod() != null))
+            var properties = typeof(JsonSerializer).GetTypeInfo().DeclaredProperties;
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/JsonSerializerExtensions.cs]: just in case, this does not appear to be related with this PR.

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Properties/Resources.resx [hunk 9]
```diff
@@ -127,6 +127,12 @@ Parameter name: {0}</value>
   <data name="ArgumentCannotBeNegative" xml:space="preserve">
     <value>The specified argument {0} cannot be initialized with a negative value.</value>
   </data>
+  <data name="AuthenticationEndpointContainsQuery" xml:space="preserve">
+    <value>The AuthenticationEndpoint must not contain a query string.</value>
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Properties/Resources.resx]: should we split with a space between "AuthenticationEndpoint"?

File: ClientRuntimes/CSharp/ClientRuntime/TokenCredentials.cs [hunk 10]
```diff
@@ -15,41 +16,52 @@ namespace Microsoft.Rest
...
+        /// <param name="token">Valid JSON Web Token (JWT).</param>
+        /// <param name="tokenType">The token type of the given token.</param>
+        public TokenCredentials(string token, string tokenType) : 
+            this(new StringTokenProvider(token, tokenType))
+        {
+             if (string.IsNullOrEmpty(token))
+            {
+                throw new ArgumentNullException("token");
+            }
+              if (string.IsNullOrEmpty(tokenType))
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime/TokenCredentials.cs]: indent

File: ClientRuntimes/CSharp/ClientRuntime.Azure/JsonSerializerExtensions.cs (hunk 8 above)
Comment (markcowl) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/JsonSerializerExtensions.cs]: This is using the .net 4.5-based reflection library instead of the .net 40 based relfection library

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs (hunk 1 above)
Comment (markcowl) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs]: I'd still like to be explicit that this is the access token, which is a string, and not the entire authentication token, which contains the refresh token and other metadata

File: ClientRuntimes/CSharp/ClientRuntime/StringTokenProvider.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        private string _accessToken;
+        private string _type;
+
+        /// <summary>
+        /// Create a token provider for bearer tokens that always returns the given 
+        /// access token.
+        /// </summary>
+        /// <param name="accessToken">The access token for this token provider to return.</param>
+        public StringTokenProvider(string accessToken) : this(accessToken, "Bearer")
```
Comment (yugangw-msft) [inline ClientRuntimes/CSharp/ClientRuntime/StringTokenProvider.cs]: The "Bearer" has been used multiple time, suggest to use a pre-defined constants

**Comment by yugangw-msft** [comment]
Please remove the "UpgradeLog.htm" and "UpgradeLog2.htm"

File: ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryUserTokenProvider.cs (hunk 3 above)
Comment (markcowl) [inline ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryUserTokenProvider.cs]: Made the class public and this constructor internal, which was the intention

**Comment by markcowl** [comment]
Done.
